ext/canvas/01_image.js

```diff
@@ -204,147 +209,73 @@ function createImageBitmap(
-
-  if (
-    sx !== undefined && sy !== undefined && sw !== undefined && sh !== undefined
-  ) {
-    sourceRectangle = [
-      [sx, sy],
-      [sx + sw, sy],
-      [sx + sw, sy + sh],
-      [sx, sy + sh],
-    ];
-  } else {
-    sourceRectangle = [
-      [0, 0],
-      [width, 0],
-      [width, height],
-      [0, height],
-    ];
-  }
-  const widthOfSourceRect = sourceRectangle[1][0] - sourceRectangle[0][0];
-  const heightOfSourceRect = sourceRectangle[3][1] - sourceRectangle[0][1];
-
-  let outputWidth;
-  if (options.resizeWidth !== undefined) {
-    outputWidth = options.resizeWidth;
-  } else if (options.resizeHeight !== undefined) {
-    outputWidth = MathCeil(
-      (widthOfSourceRect * options.resizeHeight) / heightOfSourceRect,
-    );
-  } else {
-    outputWidth = widthOfSourceRect;
-  }
-
-  let outputHeight;
-  if (options.resizeHeight !== undefined) {
-    outputHeight = options.resizeHeight;
-  } else if (options.resizeWidth !== undefined) {
-    outputHeight = MathCeil(
-      (heightOfSourceRect * options.resizeWidth) / widthOfSourceRect,
+  // 3.
+  const isBlob = ObjectPrototypeIsPrototypeOf(BlobPrototype, image);
+  const isImageData = ObjectPrototypeIsPrototypeOf(ImageDataPrototype, image);
+  if (!isBlob && !isImageData) {
+    return PromiseReject(
+      new DOMException(
+        `${prefix}: The provided value for 'image' is not of type '(${
+          ArrayPrototypeJoin(imageBitmapSources, " or ")
+        })'.`,
+        "InvalidStateError",
+      ),
     );
-  } else {
-    outputHeight = heightOfSourceRect;
   }
 
-  if (options.colorSpaceConversion === "none") {
-    throw new TypeError(
-      "Cannot create image: invalid colorSpaceConversion option, 'none' is not supported",
-    );
-  }
+  // 4.
+  return (async () => {
+    let width = 0;
+    let height = 0;
+    let mimeType = "";
+    let imageBitmapSource, buf;
+    if (isBlob) {
+      imageBitmapSource = imageBitmapSources[0];
+      buf = new Uint8Array(await image.arrayBuffer());
+      mimeType = sniffImage(image.type);
+    }
+    if (isImageData) {
+      width = image[_width];
+      height = image[_height];
+      imageBitmapSource = imageBitmapSources[1];
+      buf = new Uint8Array(TypedArrayPrototypeGetBuffer(image[_data]));
+    }
 
-  /*
-   * The cropping works differently than the spec specifies:
-   * The spec states to create an infinite surface and place the top-left corner
-   * of the image a 0,0 and crop based on sourceRectangle.
-   *
-   * We instead create a surface the size of sourceRectangle, and position
-   * the image at the correct location, which is the inverse of the x & y of
-   * sourceRectangle's top-left corner.
-   */
-  const data = op_image_process(
-    new Uint8Array(TypedArrayPrototypeGetBuffer(input)),
-    {
+    let sx;
+    if (typeof sxOrOptions === "number") {
+      sx = sxOrOptions;
+    }
+    // TODO(Hajime-san): this should be real async
+    const processedImage = op_create_image_bitmap(
```

Comment:
Initially, I thought it would be necessary to make it an async op, but when I actually implemented it that seemed to reduce performance.
In this case, is it not necessary to make it an async op?🤔

```bash
% target/release/deno bench bench.ts
    CPU | Apple M1
Runtime | Deno 2.0.0-rc.9 (aarch64-apple-darwin)

file:///path/to/deno/bench.ts

benchmark           time/iter (avg)        iter/s      (min … max)           p75      p99     p995
------------------- ----------------------------- --------------------- --------------------------
createImageBitmap            4.8 µs       208,800 (  4.7 µs …   5.1 µs)   4.8 µs   5.1 µs   5.1 µs
```